src/zerocoin_v3.cpp

```diff
@@ -167,6 +168,12 @@ bool CheckSpendZcoinTransactionV3(
 					return false;
 			}
 
+			// check duplicated serials in same transaction.
```

Comment:
This change is not necessary. We have a LARGER problem here. What we must do is move zerocoinTxInfoV3->zcTransactions.insert(hashTx); from line 182 to the end of this function. The problem we have here is the first spend will be verified by "CheckZerocoinSpendSerialV3", and we'll verify that spend is not re-used. Then "zerocoinTxInfoV3->zcTransactions.insert(hashTx);" happens, and for all subsequent spends their serials are not checked. I.E. the attacker can actually create a transaction, where the first spend is valid, and the other spends re-use the same serial, and it will pass :) Looks like we have the same problem for the old zerocoin, and we're lucky no-one used it against us.

Reply:
we ever be attacked with this bug before. and we fix this way on Zerocoin. your solution look good.  I'll try.

Reply:
@martun  If do that the transaction will be accepted to mempool but will be rejected when create a block. I think we should check in mempool level.

Reply:
@thebevrishot No, why? During adding to mempool, this line will run: zerocoinTxInfoV3->zcTransactions.insert(hashTx);

Then this transaction will not be re-checked while being added to the blocks.

Reply:
function AcceptToMemoryPoolWorker do not pass zerocoinTxInfoV3 then `zerocoinTxInfoV3->zcTransactions.insert(hashTx);` would not be call.

Reply:
@martun Was that not the same attack that occurred following the MTP launch?

Reply:
@riordant Yes, this is similiar to that attack, but I do not believe you have fixed the code properly that time.

I think even while adding the mempool we still want to check if the serial number happened before, but we can set isVerifyDB, such that serials will not be recorded as spent.

And this line still needs to get out of the for loop: zerocoinTxInfoV3->zcTransactions.insert(hashTx);

Otherwise all the other spends, except the first one will not be checked while being added to the blocks.

Reply:
for the case serial duplicate with confirmed serial is checked in function `AcceptToMemoryPoolWorker`(line: 1357). 

I agree with move `zerocoinTxInfoV3->zcTransactions.insert(hashTx);` out of loop. It do not make sense to insert single hashTx many time.

Reply:
@FreakyMaryk please create testcases that cover this kind of attack as well